snuba/clickhouse/astquery.py

```diff
@@ -0,0 +1,163 @@
+    specific customizations.
+
+    Here the process of formatting the query, is independent from
+    the query body dictionary and it is performed starting from the
+    AST.
+    """
+
+    def __init__(self, query: Query, settings: RequestSettings,) -> None:
+        # Snuba query structure
+        # Referencing them here directly since it makes it easier
+        # to process this query independently from the Snuba Query
+        # and there is no risk in doing so since they are immutable.
+        self.__selected_columns = query.get_selected_columns_from_ast()
+        self.__condition = query.get_condition_from_ast()
+        self.__groupby = query.get_groupby_from_ast()
+        self.__having = query.get_having_from_ast()
+        self.__orderby = query.get_orderby_from_ast()
+        self.__data_source = query.get_data_source()
+        self.__arrayjoin = query.get_arrayjoin_from_ast()
+        self.__granularity = query.get_granularity()
+        self.__limit = query.get_limit()
+        self.__limitby = query.get_limitby()
+        self.__offset = query.get_offset()
+
+        # Clickhouse specific fields
+        self.__prewhere = query.get_prewhere_ast()
+        self.__turbo = settings.get_turbo()
+        self.__final = query.get_final()
+
+        # Attributes that should be clickhouse specific that
+        # have to be removed from the Snuba Query
+        self.__sample = query.get_sample()
+        self.__hastotals = query.has_totals()
+
+        self.__settings = settings
+        self.__formatted_query: Optional[str] = None
+
+    def format_sql(self) -> str:
+        if self.__formatted_query:
+            return self.__formatted_query
+
+        parsing_context = ParsingContext()
+
+        selected_cols = [
+            e.accept(ClickhouseExpressionFormatter(parsing_context))
+            for e in self.__selected_columns
+        ]
+        formatted_columns = ", ".join(selected_cols)
+        select_clause = f"SELECT {formatted_columns}"
+
+        # TODO: The visitor approach will be used for the FROM clause as well.
+        from_clause = f"FROM {self.__data_source.format_from()}"
+
+        if self.__final:
+            from_clause = f"{from_clause} FINAL"
+
+        # TODO: Sampling rate will become one step of Clickhouse query processing
+        if not self.__data_source.supports_sample():
+            sample_rate = None
+        else:
+            if self.__sample:
+                sample_rate = self.__sample
+            elif self.__settings.get_turbo():
+                sample_rate = settings.TURBO_SAMPLE_RATE
+            else:
+                sample_rate = None
+        if sample_rate:
+            from_clause = f"{from_clause} SAMPLE {sample_rate}"
+
+        join_clause = ""
+        if self.__arrayjoin:
+            formatted_array_join = self.__arrayjoin.accept(
+                ClickhouseExpressionFormatter(parsing_context)
+            )
+            join_clause = f"ARRAY JOIN {formatted_array_join}"
```

Comment:
`join_clause` should probably be `array_join_clause` here, otherwise it's a bit misleading.